styx-service-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -955,20 +957,30 @@ private static StringValue jsonValue(Object o) throws JsonProcessingException {
         if (!isRetryableTransactionException(e.getCause())) {
           throw e;
         }
-        if (retrySettings.getMaxAttempts() > 0 && attempt >= retrySettings.getMaxAttempts()) {
+        checkAttemptsAndSleep(retrySettings, backoff, attempt, e);
+      } catch (CounterCapacityException e) {
```

Comment:
This method is generic handling of running transaction with retries, dealing with this specific exception by checking the error message seems like code smell. At least this should be addressed in `isRetryableTransactionException`. Or be handling at the calling site.